src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs

```diff
@@ -117,10 +119,15 @@ public TState GetRequiredState(DocumentId documentId)
     /// Get states ordered in compilation order.
     /// </summary>
     /// <returns></returns>
-    public IEnumerable<TState> GetStatesInCompilationOrder()
+    public ImmutableArray<TState> GetStatesInCompilationOrder()
     {
-        var map = States;
-        return Ids.Select(id => map[id]);
+        if (_statesInCompilationOrder.IsEmpty)
```

Comment:
Do you want this to be an .IsDefault so we aren't rerunning this if the collection is empty? It'll be cheap obviously in that case, but should still work?

Reply:
++ on this.  signing off with that change in mind.

Reply:
Will go ahead and change, I just thought it was a bit safer to always have it initialized to empty (as I don't think that's a very likely populated state for this object)